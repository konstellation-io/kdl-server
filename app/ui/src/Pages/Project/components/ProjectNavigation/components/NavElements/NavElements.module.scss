@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.projectTools {
  background-color: palette(base, 300);
  a:active > div,
  a:hover > div,
  a:focus > div {
    background-color: palette(base, 200) !important;
  }

  .userTools {
    border-top: 2px solid palette(base, 400);
    margin-top: $grid-unit;
    transition: border-right 0.3s ease;
    border-right: 2px solid transparent;

    &.started {
      border-right: 2px solid palette(success);
    }
    &.stopped {
      border-right: 2px solid palette(alert);
    }
    &.loading {
      border-right: 2px solid palette(feedback);
    }

    .userToolLabel {
      @include font-caption;
      display: flex;
      color: font-color(regular);
      width: 100%;
      white-space: nowrap;
      margin-left: 2 * $grid-unit;
      @include circle-before(#3d3d3d);
      &.pending:before {
        background-color: palette(alert);
        top: $grid-unit - 5px;
      }
      &.running:before {
        background-color: palette(success);
        top: $grid-unit - 5px;
      }
      &.error:before {
        background-color: palette(error);
        top: $grid-unit - 5px;
      }
      padding-left: 3 * $grid-unit;
      flex-direction: row;
      justify-content: flex-start;
      align-items: center;
    }

    .usertoolsIcon {
      margin: $grid-unit;
      cursor: pointer;
      &.blocked {
        color: palette(alert);
      }
      &:active,
      &:hover,
      &:focus {
        color: palette(highlight);
      }
    }

    .usertoolsOptions {
      display: flex;
      align-items: center;
      justify-content: space-between;

      div[data-for="settings"],
      div[data-for="start"],
      div[data-for="stop"] {
        display: flex;
        justify-content: center;
        svg {
          width:  $grid-unit * 2;
          height: $grid-unit * 2;
        }
      }
      &.opened {
        height: 40px;
        div[data-for="settings"],
        div[data-for="start"],
        div[data-for="stop"] {
          svg {
            width: $grid-unit + 4;
            height: $grid-unit + 4;
          }
        }
      }
    }

    .usertoolsSettings {
      display: flex;
      flex-direction: column;
      width: 100%;
      .progressSpinnerContainer {
        margin-top: 0;
        margin-left: 13px;
      }
      &.opened {
        flex-direction: row;
        width: auto;
        .progressSpinnerContainer {
          margin-top: $grid-unit;
          margin-left: $grid-unit;
          margin-right: $grid-unit;
        }
      }
    }

    .toggleToolsWrapper {
      border-top: 2px dotted palette(base, 400);
      transition: opacity 0.3s ease;
      .loadingTools {
        color: palette(highlight);
        margin: 0 1.4 * $grid-unit;
        transition: transform ease 0.3s;
        flex-shrink: 0;
      }
      &.disabled {
        pointer-events: none;
        opacity: 0.5;
      }
    }
  }
}
