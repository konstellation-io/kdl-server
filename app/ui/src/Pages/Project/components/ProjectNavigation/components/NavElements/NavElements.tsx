import AnimateHeight from 'react-animate-height';
import CircularProgress from '@material-ui/core/CircularProgress';
import {NavButtonLink} from '../../ProjectNavigation';
import NavigationButton from '../NavigationButton/NavigationButton';
import IconPause from '@material-ui/icons/Pause';
import IconPlay from '@material-ui/icons/PlayArrow';
import IconSettings from '@material-ui/icons/Settings';
import * as React from 'react';
import {RouteProjectParams} from 'Constants/routes';
import cx from 'classnames';
import styles from './NavElements.module.scss';
import {useParams} from 'react-router-dom';
import useProjectNavigation from 'Hooks/useProjectNavigation';
import {useReactiveVar} from '@apollo/client';
import {lastRanRuntime, loadingRuntime, primaryPanel, runningRuntime} from 'Graphql/client/cache';
import usePanel, {PanelType} from 'Graphql/client/hooks/usePanel';
import {USERTOOLS_PANEL_OPTIONS} from 'Pages/Project/panelSettings';
import {PANEL_ID} from 'Graphql/client/models/Panel';
import RuntimeRunner, {RuntimeAction} from 'Components/RuntimeRunner/RuntimeRunner';
import Tooltip from 'Components/Tooltip/Tooltip';

type Props = {
  isOpened: boolean;
};

function NavElements({ isOpened }: Props) {
  const { projectId } = useParams<RouteProjectParams>();
  const { mainRoutes, userToolsRoutes, projectToolsRoutes } = useProjectNavigation(projectId);
  const runtimeLoading = useReactiveVar(loadingRuntime);
  const isLoading = runtimeLoading !== null;

  const runtimeRunning = useReactiveVar(runningRuntime);
  const panelData = useReactiveVar(primaryPanel);
  const runtimeLastRun = useReactiveVar(lastRanRuntime);

  const { openPanel: openRuntimesList, closePanel: closeRuntimesList } = usePanel(
    PanelType.PRIMARY,
    USERTOOLS_PANEL_OPTIONS,
  );

  const tooltipProps = {
    effect: 'solid',
    textColor: 'white',
    backgroundColor: '#888',
  };

  function toggleUsertoolsPanel() {
    const shouldOpen = !panelData || panelData.id !== PANEL_ID.RUNTIMES_LIST;

    if (shouldOpen) openRuntimesList();
    else closeRuntimesList();
  }

  function renderToggleToolsIcon() {
    const Progress = (
      <div className={styles.progressSpinnerContainer}>
        <CircularProgress color="inherit" className={styles.loadingTools} size={12} />{' '}
      </div>
    );
    if (isLoading) return Progress;

    return runtimeRunning ? (
      <Tooltip tooltipId="stop" spanText="Stop tools" tooltipProps={tooltipProps}>
        <RuntimeRunner action={RuntimeAction.Stop}>
          <IconPause className={cx(styles.usertoolsIcon, 'icon-small')} data-testid="stopTools" />
        </RuntimeRunner>
      </Tooltip>
    ) : (
      <Tooltip tooltipId="start" spanText="Start tools" tooltipProps={tooltipProps}>
        <RuntimeRunner action={RuntimeAction.Start} runtime={runtimeLastRun || undefined}>
          <IconPlay className={cx(styles.usertoolsIcon, 'icon-small')} data-testid="startTools" />
        </RuntimeRunner>
      </Tooltip>
    );
  }

  return (
    <>
      {mainRoutes.map(({ Icon, label, route }) => (
        <NavButtonLink to={route} key={label}>
          <NavigationButton label={label} Icon={Icon} isNavCollapsed={!isOpened} />
        </NavButtonLink>
      ))}
      <div className={styles.projectTools}>
        {projectToolsRoutes.map(({ Icon, label, route, disabled }) => (
          <NavButtonLink to={route} key={label} disabled={disabled}>
            <NavigationButton label={label} Icon={Icon} isNavCollapsed={!isOpened} />
          </NavButtonLink>
        ))}
        <div
          className={cx(styles.userTools, {
            [styles.started]: runtimeRunning,
            [styles.stopped]: !runtimeRunning,
            [styles.loading]: isLoading,
          })}
        >
          <div className={cx(styles.usertoolsOptions, { [styles.opened]: isOpened })}>
            <AnimateHeight height={isOpened ? 'auto' : 0} duration={300}>
              <div className={styles.userToolLabel}>USER TOOLS</div>
            </AnimateHeight>
            <div
              className={cx(styles.usertoolsSettings, { [styles.opened]: isOpened })}
              data-testid="usertoolsSettings"
            >
              <Tooltip tooltipId="settings" spanText="Show available runtimes" tooltipProps={tooltipProps}>
                <div data-tip data-for="settings" data-testid="openRuntimeSettings">
                  <IconSettings className={cx(styles.usertoolsIcon, 'icon-small')} onClick={toggleUsertoolsPanel} />
                </div>
              </Tooltip>
              {renderToggleToolsIcon()}
            </div>
          </div>
          {userToolsRoutes.map(({ Icon, label, route, disabled }) => (
            <NavButtonLink to={route} key={label} disabled={disabled}>
              <NavigationButton label={label} Icon={Icon} isNavCollapsed={!isOpened} />
            </NavButtonLink>
          ))}
          <div
            className={cx(styles.toggleToolsWrapper, {
              [styles.disabled]: isLoading,
            })}
            data-testid="confirmationModal"
          />
        </div>
      </div>
    </>
  );
}

export default NavElements;
