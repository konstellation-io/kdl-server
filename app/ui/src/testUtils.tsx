import * as React from 'react';
import { ApolloClient, ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { mount } from 'enzyme';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function apolloRender(component: JSX.Element, client: ApolloClient<any>) {
  render(<ApolloProvider client={client}>{component}</ApolloProvider>);
}

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export const HookWrapper = ({ hook }: any) => <div hook={hook()} />; // eslint-disable-line @typescript-eslint/no-explicit-any
export async function apolloHookRender(
  // eslint-disable-next-line @typescript-eslint/ban-types
  hook: Function,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  client: ApolloClient<any>,
) {
  const wrapper = mount(
    <ApolloProvider client={client}>
      <HookWrapper hook={hook} />
    </ApolloProvider>,
  );

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return wrapper.find('div').props().hook;
}

export const loadingHandler = () =>
  new Promise(() => {
    // This is intentional
  });
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const dataHandler = (data: any) => () => Promise.resolve({ data });
export const errorHandler = (error: string) => () => Promise.reject({ error });

export function getSnapshot() {
  return document.querySelector('body');
}

export const ERROR_MESSAGE = 'Something went wrong, refresh the page or try again later.';
