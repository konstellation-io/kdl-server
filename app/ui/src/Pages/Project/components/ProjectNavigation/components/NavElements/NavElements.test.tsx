import * as React from 'react';
import NavElements from './NavElements';
import GetMeQuery from 'Graphql/queries/getMe';
import SetActiveUserToolsMutation from 'Graphql/mutations/setActiveUserTools';
import { createMockClient } from 'mock-apollo-client';
import { screen, fireEvent, cleanup, within } from '@testing-library/react';
import { apolloRender, loadingHandler, dataHandler, getSnapshot } from 'testUtils';
import data from 'Mocks/GetMeQuery';
import mutationData from 'Mocks/SetActiveProjectToolsMutation';
import { StaticRouter } from 'react-router';

const Component = (
  <StaticRouter>
    <NavElements isOpened />
  </StaticRouter>
);
let mockClient;
beforeEach(() => {
  mockClient = createMockClient();
});
afterEach(cleanup);

describe('When component is loading', () => {
  beforeEach(() => {
    mockClient.setRequestHandler(GetMeQuery, loadingHandler);
    apolloRender(Component, mockClient);
  });

  it('should render without crashing', () => {
    expect(screen.queryByText('tools')).not.toBeInTheDocument();
  });
});

describe('When data is ready', () => {
  beforeEach(() => {
    mockClient.setRequestHandler(GetMeQuery, dataHandler(data));
    apolloRender(Component, mockClient);
  });

  it('should render without crashing', () => {
    expect(getSnapshot()).toMatchSnapshot();
  });

  it('should show confirmation modal when stopping user tools', () => {
    // Act.
    const stopToolButton = screen.getByText('stop tools', { exact: false });
    fireEvent.click(stopToolButton);

    // Assert.
    const modal = screen.getByText('STOP YOUR TOOLS', { exact: false });
    expect(modal).toBeInTheDocument();
  });
});

describe('When stopping tools', () => {
  const mockMutation = jest.fn();

  beforeEach(() => {
    mockMutation.mockResolvedValue({ data: mutationData });

    mockClient.setRequestHandler(GetMeQuery, dataHandler(data));
    mockClient.setRequestHandler(SetActiveUserToolsMutation, mockMutation);
    apolloRender(Component, mockClient);
  });

  afterEach(() => {
    mockMutation.mockClear();
  });

  it('should call mutation function', () => {
    // Act.
    const stopToolButton = screen.getByText('stop tools', { exact: false });
    fireEvent.click(stopToolButton);

    const modal = screen.getByTestId('confirmationModal');
    const acceptButton = within(modal).getByText('Stop tools', {
      exact: false,
    });
    fireEvent.click(acceptButton);

    // Assert.
    setTimeout(() => {
      expect(mockMutation).toHaveBeenCalled();
    }, 0);
  });

  it('should not call mutation function when cancelling the operation', () => {
    // Act.
    const stopToolButton = screen.getByText('stop tools', { exact: false });
    fireEvent.click(stopToolButton);

    const modal = screen.getByTestId('confirmationModal');
    const cancelButton = within(modal).getByText('Cancel', {
      exact: false,
    });
    fireEvent.click(cancelButton);

    // Assert.
    setTimeout(() => {
      expect(mockMutation).not.toHaveBeenCalled();
    }, 0);
  });
});
