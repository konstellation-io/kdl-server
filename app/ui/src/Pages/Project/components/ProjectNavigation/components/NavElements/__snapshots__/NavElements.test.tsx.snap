// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When data is ready should render without crashing 1`] = `
<body>
  <div>
    <a
      class=""
      href="/projects/undefined/overview"
    >
      <div
        class="navButton"
        title="Overview"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon-regular icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 13h8V3H3v10zm0 8h8v-6H3v6zm10 0h8V11h-8v10zm0-18v6h8V3h-8z"
          />
        </svg>
        <span
          class="label"
        >
          Overview
        </span>
      </div>
    </a>
    <div
      class="projectTools"
    >
      <a
        class=""
        href="/projects/undefined/tool/kg"
      >
        <div
          class="navButton"
          title="Knowledge Galaxy"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root icon-regular icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 3c-.46 0-.93.04-1.4.14-2.76.53-4.96 2.76-5.48 5.52-.48 2.61.48 5.01 2.22 6.56.43.38.66.91.66 1.47V19c0 1.1.9 2 2 2h.28c.35.6.98 1 1.72 1s1.38-.4 1.72-1H14c1.1 0 2-.9 2-2v-2.31c0-.55.22-1.09.64-1.46C18.09 13.95 19 12.08 19 10c0-3.87-3.13-7-7-7zm2 16h-4v-1h4v1zm0-2h-4v-1h4v1zm-1.5-5.59V14h-1v-2.59L9.67 9.59l.71-.71L12 10.5l1.62-1.62.71.71-1.83 1.82z"
            />
          </svg>
          <span
            class="label"
          >
            Knowledge Galaxy
          </span>
        </div>
      </a>
      <a
        class=""
        href="/projects/undefined/tool/gitea"
      >
        <div
          class="navButton"
          title="Gitea"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root icon-regular icon"
            focusable="false"
            viewBox="5.370000000000001 33.59 133.015 81.32399999999998"
          >
            <path
              d="M27.71 33.618c-9.547-.028-22.34 6.798-21.632 23.904C7.184 84.25 31.535 86.73 41.27 86.945c1.068 5.013 12.522 22.305 21.002 23.216h37.152c22.278-1.668 38.961-75.756 26.594-76.038-46.782 2.477-49.996 2.138-88.6 0-2.495-.027-5.972-.495-9.708-.505zm2.491 9.458C31.553 56.77 33.757 64.78 38.22 77.02c-11.382-1.505-21.07-5.225-22.851-19.11-.951-7.411 2.39-15.168 14.833-14.834z"
            />
          </svg>
          <span
            class="label"
          >
            Gitea
          </span>
        </div>
      </a>
      <a
        class=""
        href="/projects/undefined/tool/drone"
      >
        <div
          class="navButton"
          title="Drone"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root icon-regular icon"
            focusable="false"
            viewBox="0 0 2500 2500"
          >
            <path
              clip-rule="evenodd"
              d="m503.6 242.3-10.7 10.7 438.1 438.1c-62 96.7-97.6 215.5-97.6 350.6 0 375.6 275.8 625 625 625 130.5 0 250.8-34.8 350.2-98l440.3 440.3c-228.1 301.3-590.7 491-998.8 491-690.3 0-1250.1-542.6-1250.1-1250 0-420.7 198-783.1 503.6-1007.7zm208.4-123.8c162.9-76.2 345.4-118.5 538-118.5 690.3 0 1250 542.6 1250 1250 0 194.6-42.3 376.7-118.1 538.5l-396.2-396.2c62-96.7 97.6-215.5 97.6-350.6 0-375.6-275.8-625-625-625-130.5 0-250.8 34.8-350.2 98zm746.3 1298.2c-209.6 0-375-149.6-375-375s165.4-375 375-375 375 149.6 375 375-165.4 375-375 375z"
              fill-rule="evenodd"
            />
          </svg>
          <span
            class="label"
          >
            Drone
          </span>
        </div>
      </a>
      <a
        class=""
        href="/projects/undefined/tool/filebrowser"
      >
        <div
          class="navButton"
          title="Filebrowser"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root icon-regular icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
            />
          </svg>
          <span
            class="label"
          >
            Filebrowser
          </span>
        </div>
      </a>
      <a
        class=""
        href="/projects/undefined/tool/mlflow"
      >
        <div
          class="navButton"
          title="Mlflow"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root icon-regular icon"
            focusable="false"
            viewBox="0 0 24 25"
          >
            <path
              d="M18.9509 2.17634C14.0635 -1.24345 7.40926 -0.571051 3.30485 3.75735C-0.799556 8.08576 -1.11776 14.7663 2.55667 19.4652L6.22524 16.7724C4.40423 14.5134 4.03437 11.4124 5.27298 8.78843C6.51159 6.16448 9.1408 4.47914 12.0422 4.44929L11.9554 7.31581L18.9509 2.17634Z"
            />
            <path
              d="M21.6639 4.85176C21.5423 4.68382 21.4149 4.51878 21.2846 4.35953L17.7521 6.96546C19.7463 9.17232 20.2637 12.3424 19.0746 15.0688C17.8856 17.7952 15.2104 19.5728 12.2362 19.6129L12.3231 16.7493L5.24945 21.9611C10.0971 25.2255 16.5896 24.5337 20.6406 20.3213C24.6917 16.1089 25.1295 9.59437 21.6784 4.87782L21.6639 4.85176Z"
            />
          </svg>
          <span
            class="label"
          >
            Mlflow
          </span>
        </div>
      </a>
      <div
        class="userTools started"
      >
        <div
          aria-hidden="false"
          class="rah-static rah-static--height-auto"
          style="height: auto; overflow: visible;"
        >
          <div>
            <div
              class="userToolLabel"
            >
              USER TOOLS
            </div>
          </div>
        </div>
        <a
          class=""
          href="/projects/undefined/tool/jupyter"
        >
          <div
            class="navButton"
            title="Jupyter"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root icon-regular icon"
              focusable="false"
              viewBox="0 0 44 51"
            >
              <g>
                <path
                  d="M 6.03059 2.83565C 6.06715 3.43376 5.92485 4.02921 5.6218 4.54615C 5.31875 5.0631 4.86869 5.47813 4.32893 5.73839C 3.78917 5.99864 3.18416 6.09233 2.59097 6.00753C 1.99778 5.92272 1.44326 5.66326 0.998048 5.26219C 0.552837 4.86113 0.23709 4.33661 0.0910307 3.75546C -0.0550287 3.17431 -0.0247891 2.56283 0.177897 1.99893C 0.380583 1.43503 0.746541 0.944221 1.22915 0.589037C 1.71176 0.233853 2.28918 0.0303686 2.88784 0.00450543C 3.28035 -0.0170932 3.67326 0.0391144 4.04396 0.169896C 4.41467 0.300677 4.75587 0.503453 5.04794 0.766561C 5.34 1.02967 5.57718 1.34792 5.74582 1.70301C 5.91446 2.0581 6.01124 2.44303 6.03059 2.83565L 6.03059 2.83565Z"
                  transform="translate(33.48 0.69)"
                />
                <path
                  d="M 18.6962 7.12238C 10.6836 7.12238 3.64131 4.24672 0 0C 1.41284 3.82041 3.96215 7.1163 7.30479 9.44404C 10.6474 11.7718 14.623 13.0196 18.6962 13.0196C 22.7695 13.0196 26.745 11.7718 30.0877 9.44404C 33.4303 7.1163 35.9796 3.82041 37.3925 4.0486e-13C 33.7601 4.24672 26.7445 7.12238 18.6962 7.12238Z"
                  transform="translate(3.21 31.27)"
                />
                <path
                  d="M 18.6962 5.89725C 26.7089 5.89725 33.7512 8.77291 37.3925 13.0196C 35.9796 9.19922 33.4303 5.90333 30.0877 3.57559C 26.745 1.24785 22.7695 4.0486e-13 18.6962 0C 14.623 4.0486e-13 10.6474 1.24785 7.30479 3.57559C 3.96215 5.90333 1.41284 9.19922 0 13.0196C 3.64131 8.76401 10.648 5.89725 18.6962 5.89725Z"
                  transform="translate(3.21 4.88)"
                />
                <path
                  d="M 7.59576 3.56656C 7.64276 4.31992 7.46442 5.07022 7.08347 5.72186C 6.70251 6.3735 6.13619 6.89698 5.45666 7.22561C 4.77713 7.55424 4.01515 7.67314 3.26781 7.56716C 2.52046 7.46117 1.82158 7.13511 1.26021 6.63051C 0.698839 6.12591 0.300394 5.46561 0.115637 4.73375C -0.0691191 4.00188 -0.0318219 3.23159 0.222777 2.52099C 0.477376 1.8104 0.93775 1.19169 1.54524 0.743685C 2.15274 0.295678 2.87985 0.0386595 3.63394 0.00537589C 4.12793 -0.0210471 4.62229 0.0501173 5.08878 0.214803C 5.55526 0.37949 5.98473 0.63447 6.35264 0.965179C 6.72055 1.29589 7.01971 1.69584 7.233 2.1422C 7.4463 2.58855 7.56957 3.07256 7.59576 3.56656L 7.59576 3.56656Z"
                  transform="translate(3.28 43.09)"
                />
                <path
                  d="M 2.25061 4.37943C 1.81886 4.39135 1.39322 4.27535 1.02722 4.04602C 0.661224 3.81668 0.371206 3.48424 0.193641 3.09052C 0.0160762 2.69679 -0.0411078 2.25935 0.0292804 1.83321C 0.0996686 1.40707 0.294486 1.01125 0.589233 0.695542C 0.883981 0.37983 1.2655 0.158316 1.68581 0.0588577C 2.10611 -0.0406005 2.54644 -0.0135622 2.95143 0.136572C 3.35641 0.286707 3.70796 0.553234 3.96186 0.902636C 4.21577 1.25204 4.3607 1.66872 4.37842 2.10027C 4.39529 2.6838 4.18131 3.25044 3.78293 3.67715C 3.38455 4.10387 2.83392 4.35623 2.25061 4.37943Z"
                  transform="translate(1.87 5.43)"
                />
              </g>
            </svg>
            <span
              class="label"
            >
              Jupyter
            </span>
          </div>
        </a>
        <a
          class=""
          href="/projects/undefined/tool/vscode"
        >
          <div
            class="navButton"
            title="Vscode"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root icon-regular icon"
              focusable="false"
              viewBox="0 0 234 235.1"
            >
              <path
                class="st0"
                d="M83.3 143.9l-58 45.2L0 176.5V58.7L25.2 46l57.6 45.3L174 0l60 23.9v186.9l-59.7 24.3-91-91.2zm88.9 15.9V75.3l-54.6 42.3 54.6 42.2zM27.3 144.6L56 118.5 27.3 89.9v54.7z"
              />
            </svg>
            <span
              class="label"
            >
              Vscode
            </span>
          </div>
        </a>
        <div
          class="toggleToolsWrapper"
          data-testid="confirmationModal"
        >
          <div>
            <div
              class="navButton"
              data-testid="stopTools"
              title="Stop tools"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root icon-regular icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19h4V5H6v14zm8-14v14h4V5h-4z"
                />
              </svg>
              <span
                class="label"
              >
                Stop tools
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
